Index: mod_auth_mysql-3.0.0/mod_auth_mysql.c
===================================================================
--- mod_auth_mysql-3.0.0.orig/mod_auth_mysql.c
+++ mod_auth_mysql-3.0.0/mod_auth_mysql.c
@@ -1539,6 +1539,7 @@
 	char *auth_table = "mysql_auth", *auth_user_field = "username",
 		*auth_password_field = "passwd", *auth_password_clause = "";
 	char *query;
+    char *salt_and_passwordhash;
 	char *esc_user = NULL;
 	MYSQL_RES *result;
 	MYSQL_ROW sql_row;
@@ -1578,7 +1579,7 @@
 							, auth_table, esc_user,auth_password_clause);
 
 	query = (char *) PSTRCAT(r->pool, "SELECT ", auth_password_field,
-					" FROM ", auth_table, " WHERE ",
+					", salt FROM ", auth_table, " WHERE ",
 					auth_user_field, "='", esc_user, "'",
 					auth_password_clause, NULL);
 	if (!query) {
@@ -1617,8 +1618,10 @@
 					"No row returned or NULL value: %s", mysql_error(sec->dbh));
 				return -1;
 			}
-			
-			rv = check_password(password, sql_row[0], r, sec);
+
+			salt_and_passwordhash = (char *) PSTRCAT(r->pool, sql_row[1], sha1_hex_hash(password), NULL);
+            APACHELOG(APLOG_DEBUG, r, "salt: [%s]", salt_and_passwordhash);
+            rv = check_password(salt_and_passwordhash, sql_row[0], r, sec); 
 			if (rv == 0)
 			{
 				APACHELOG(APLOG_INFO, r,
