--- xbmc-rpi-2acdae7/configure.in.orig	2012-07-23 16:18:49.843413933 +0200
+++ xbmc-rpi-2acdae7/configure.in	2012-07-23 16:27:46.632627764 +0200
@@ -583,16 +583,13 @@
      target_platform=target_raspberry_pi
      use_neon=no
      use_texturepacker=yes
-     use_texturepacker_native=yes
      use_arch="arm"
      use_cpu=arm1176jzf-s
-     use_armv5te="yes"
+     use_armv5te="no"
      use_armv6t2="no"
      use_armv6="yes"
      use_armvfp="yes"
      use_hardcoded_tables="yes"
-     use_alsa="yes"
-     USE_TEXTUREPACKER_NATIVE_ROOT="$TEXTUREPACKER_NATIVE_ROOT"
      ARCH="arm"
      AC_SUBST(ARCH_DEFINES, "-DTARGET_POSIX -DTARGET_LINUX -D_LINUX -D_ARMEL -DTARGET_RASPBERRY_PI")
      AC_DEFINE(HAS_EGLGLES, [1], [Define if supporting EGL based GLES Framebuffer])
@@ -600,6 +597,8 @@
      USE_OMXLIB=1; AC_DEFINE([HAVE_OMXLIB],[1],["Define to 1 if OMX libs is enabled"])
      CFLAGS="$CFLAGS"
      CXXFLAGS="$CXXFLAGS"
+     CPPFLAGS="-I/opt/vc/include $CPPFLAGS"
+     LDFLAGS="-L/opt/vc/lib $LDFLAGS"
      ;;
   *)
      use_alsa="yes"
@@ -758,10 +757,11 @@
     AC_MSG_RESULT(== WARNING: OpenGLES support is assumed.)
   else
     if test "$target_platform" = "target_raspberry_pi"; then
-      AC_DEFINE([HAVE_LIBEGL],[1],["Define to 1 if you have the `EGL' library (-lEGL)."])
-      AC_DEFINE([HAVE_LIBGLESV2],[1],["Define to 1 if you have the `GLESv2' library (-lGLESv2)."])
-      AC_MSG_RESULT(== WARNING: OpenGLES support is assumed.)
-      LIBS="$LIBS -lEGL -lGLESv2 -lbcm_host -lvcos -lvchiq_arm"
+      AC_CHECK_LIB([EGL],   [main],, AC_MSG_ERROR($missing_library))
+      AC_CHECK_LIB([GLESv2],[main],, AC_MSG_ERROR($missing_library))
+      AC_CHECK_HEADER([EGL/egl.h],, AC_MSG_ERROR($missing_library))
+      AC_CHECK_HEADER([GLES2/gl2.h],, AC_MSG_ERROR($missing_library))
+      LIBS="$LIBS -lbcm_host -lvcos -lvchiq_arm"
     else
       AC_CHECK_LIB([EGL],   [main],, AC_MSG_ERROR($missing_library))
       AC_CHECK_LIB([GLESv2],[main],, AC_MSG_ERROR($missing_library))
