diff -urN raspberrypi-tools-9999.orig/mkimage/imagetool-uncompressed.py raspberrypi-tools-9999/mkimage/imagetool-uncompressed.py
--- raspberrypi-tools-9999.orig/mkimage/imagetool-uncompressed.py	2012-06-08 01:40:56.101199697 +0200
+++ raspberrypi-tools-9999/mkimage/imagetool-uncompressed.py	2012-06-08 01:50:18.250990301 +0200
@@ -5,41 +5,46 @@
 import sys
 
 try:
-   kernel_image = sys.argv[1]
+    kernel_image = sys.argv[1]
 except:
-   kernel_image = ""
+    kernel_image = None
+
+if kernel_image is None:
+    print("usage : %s <kernel image> [kernel installation filename]" % sys.argv[0]);
+    sys.exit(0)
+
+try:
+    kernel_dest = sys.argv[2]
+except:
+    kernel_dest = "kernel.img"
 
-if kernel_image == "":
-  print("usage : imagetool-uncompressed.py <kernel image>");
-  sys.exit(0)
-   
 re_line = re.compile(r"0x(?P<value>[0-9a-f]{8})")
 
 mem = [0 for i in range(32768)]
 
 def load_to_mem(name, addr):
-   f = open(name)
+    f = open(name)
 
-   for l in f.readlines():
-      m = re_line.match(l)
+    for l in f.readlines():
+         m = re_line.match(l)
 
-      if m:
-         value = int(m.group("value"), 16)
+         if m:
+             value = int(m.group("value"), 16)
 
-         for i in range(4):
-            mem[addr] = int(value >> i * 8 & 0xff)
-            addr += 1
+             for i in range(4):
+                 mem[addr] = int(value >> i * 8 & 0xff)
+                 addr += 1
 
-   f.close()
+    f.close()
 
-load_to_mem("boot-uncompressed.txt", 0x00000000)
-load_to_mem("args-uncompressed.txt", 0x00000100)
+load_to_mem("/usr/share/raspberrypi-tools/mkimage/boot-uncompressed.txt", 0x00000000)
+load_to_mem("/usr/share/raspberrypi-tools/mkimage/args-uncompressed.txt", 0x00000100)
 
 f = open("first32k.bin", "wb")
 
 for m in mem:
-   f.write(chr(m))
+    f.write(chr(m))
 
 f.close()
 
-os.system("cat first32k.bin " + kernel_image + " > kernel.img")
+os.system("cat first32k.bin " + kernel_image + " > %s" % kernel_dest)
